lib/cabalProjectToNix.nix

```diff
@@ -17,13 +18,72 @@ let
         type == "directory" ||
         pkgs.lib.any (i: (pkgs.lib.hasSuffix i path)) [ ".project" ".cabal" "package.yaml" ];
     };
+
+  span = pred: list:
+    let n = pkgs.lib.lists.foldr (x: acc: if pred x then acc + 1 else 0) 0 list;
+    in { fst = pkgs.lib.lists.take n list; snd = pkgs.lib.lists.drop n list; };
+
+  # Parse lines of a source-repository-package block
+  parseBlockLines = blockLines: builtins.listToAttrs (builtins.map (s:
+    let pair = builtins.match " *([^:]*): *(.*)" s;
+    in pkgs.lib.attrsets.nameValuePair
+          (builtins.head pair)
+          (builtins.elemAt pair 1)) blockLines);
+
+  fetchRepo = repo: (pkgs.fetchgit {
+    url = repo.location;
+    rev = repo.tag;
+    sha256 = repo."--sha256";
```

Comment:
Initially, I thought "surely fetchgit is okay if you supply a rev?". Weirdly, it's allowed with `--pure-eval`, but not with `--restrict-eval` :thinking: Maybe this is a bug? It seems like `--pure-eval` has a more sensible notion of what "pure" things are.

Reply:
(I can now see the need for this trick, but it makes me a bit sad)

Reply:
It might actually be a nice feature for `cabal` if it could check the `sha256` of `source-repository-package`.  Perhaps less important when the `location` is `https`, but when it is `http` it would probably be nice.  It would also make sense if/when `cabal` `source-repository-package` gets tarball support.

Reply:
Shouldn't the git sha be enough? Definitely useful for tarball support, I agree.

Reply:
> Shouldn't the git sha be enough? Definitely useful for tarball support, I agree.

I'm not sure why, but `nixpkgs.fetchgit` does not support using just the git sha.  I don't think that is likely to change soon.